src/java/org/apache/cassandra/io/sstable/format/SSTableReaderBuilder.java

```diff
@@ -395,12 +393,13 @@ void load(boolean recreateBloomFilter,
                   StatsMetadata statsMetadata,
                   Set<Component> components) throws IOException
         {
-            try (FileHandle.Builder ibuilder = new FileHandle.Builder(descriptor.filenameFor(Component.PRIMARY_INDEX))
-                                               .mmapped(DatabaseDescriptor.getIndexAccessMode() == Config.DiskAccessMode.mmap)
-                                               .withChunkCache(ChunkCache.instance);
-                 FileHandle.Builder dbuilder = new FileHandle.Builder(descriptor.filenameFor(Component.DATA)).mmapped(DatabaseDescriptor.getDiskAccessMode() == Config.DiskAccessMode.mmap)
-                                                                                                             .withChunkCache(ChunkCache.instance))
+            try
             {
+                FileHandle.Builder ibuilder = new FileHandle.Builder(descriptor.fileFor(Component.PRIMARY_INDEX))
```

Comment:
Is everything that can go wrong / throw an exception covered in the `complete` block? Can we leave mmap regions or memory (e.g. compression offsets) unreleased?

Reply:
`FileHandleBuilder` being closeable always annoyed me; I do prefer to make it a trivial builder that cannot fail or that doesn't manage any state. However, we will have to manage that state explicitly in its users now.

Reply:
hmm... as I'm looking at this now, I think it needs some fixes - channel proxy is not closed in case of a failure. I also missed `CompressionMetadata`... it is  not closeable. I think maybe could implement `SharedCloseable`?

Reply:
I can do that, though I'm pretty convenient about safety of the current approach; On the other hand, I must agree with you that it will be more consistent and elegant to move the management of that outside.

Reply:
What do you think about instead of managing it in a builder, creating a method in `FileHandle`, something like `duplicateWithExtendedLength(CompressionMetadata, long)`, which retrieves regions from the current rebufferer if applicable (or we could keep mmapped  regions as a field in `FileHandle`)?

This way, it should be super clear - we create a copy of regions when we really copy a file handle

Reply:
The duplication idea sounds good to me.